src/server/router/router.go

```diff
@@ -0,0 +1,111 @@
+// Copyright Project Harbor Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package router
+
+import (
+	"context"
+	"errors"
+	"github.com/astaxie/beego"
+	beegocontext "github.com/astaxie/beego/context"
+	"github.com/goharbor/harbor/src/server/middleware"
+	"net/http"
+)
+
+type contextKeyInput struct{}
+
+// NewRoute creates a new route
+func NewRoute() *Route {
+	return &Route{}
+}
+
+// Route stores the information that matches a request
+type Route struct {
+	methods     []string
+	path        string
+	middlewares []middleware.Middleware
+}
+
+// Methods sets the methods that the route matches
+func (r *Route) Methods(methods ...string) *Route {
```

Comment:
As you have a method chain way, why need to pass method array here? 

`Method(m string)` seems is enough.

Reply:
`WithMethod` is more suitable as it is op.